File: app/src/main/java/ru/geekbrains/notes/view/NotesFragment.java
package ru.geekbrains.notes.view;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.res.Configuration;
import android.os.Bundle;
import android.util.Log;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Objects;

import ru.geekbrains.notes.MainActivity;
import ru.geekbrains.notes.Note;
import ru.geekbrains.notes.R;
import ru.geekbrains.notes.data.CardData;
import ru.geekbrains.notes.data.CardSource;
import ru.geekbrains.notes.data.CardSourceRemoteImplementation;
import ru.geekbrains.notes.data.CardsSourceResponse;
import ru.geekbrains.notes.observation.Observer;
import ru.geekbrains.notes.observation.Publisher;

public class NotesFragment extends Fragment {

    Note currentNote;
    boolean isLandscape;
    private CardSource data;
    private NotesAdapter adapter;
    private RecyclerView recyclerView;
    private Navigation navigation;
    private Publisher publisher;

    public static NotesFragment newInstance() { return new NotesFragment(); }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        setHasOptionsMenu(true);
        View view = inflater.inflate(R.layout.fragment_notes, container, false);

        this.recyclerView = view.findViewById(R.id.recyclerView);
        recyclerView.setHasFixedSize(true);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
        recyclerView.setLayoutManager(linearLayoutManager);

        this.adapter = new NotesAdapter(this);

        adapter.setNotesOnClickListener(new NotesOnClickListener() {
            @Override
            public void onNoteClick(View view, int position) {
                currentNote = new Note(data.getCardData(position).getTitle(),
                        data.getCardData(position).getText());
                isLandscape = getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE;
                showNote();
            }
        });
        recyclerView.setAdapter(adapter);

        DefaultItemAnimator defaultItemAnimator = new DefaultItemAnimator();
        defaultItemAnimator.setChangeDuration(1000);
        defaultItemAnimator.setRemoveDuration(1000);
        recyclerView.setItemAnimator(defaultItemAnimator);

        this.data = new CardSourceRemoteImplementation().init(new CardsSourceResponse() {
            @Override
            public void initialized(CardSource cardSource) {
                adapter.notifyDataSetChanged();
            }
        });
        adapter.setDataSource(data);

        return view;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        MainActivity activity = (MainActivity) context;
        this.navigation = activity.getNavigation();
        this.publisher = activity.getPublisher();
    }

    @Override
    public void onDetach() {
        this.navigation = null;
        this.publisher = null;
        super.onDetach();
    }

    private void showNote() {
        if (isLandscape) {
            showNoteLand();
        } else {
            showNotePort();
        }
    }

    private void showNotePort() {
        requireActivity()
                .getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.notes_container, NoteFragment.newInstance(currentNote))
                .addToBackStack("")
                .commit();
    }

    private void showNoteLand() {
        requireActivity()
                .getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.note_container, NoteFragment.newInstance(currentNote))
                .commit();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.fragment_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_add:
                this.navigation.addFragment(CardUpdateFragment.newInstance(), true);
                this.publisher.subscribe(new Observer() {
                    @Override
                    public void updateState(CardData cardData) {
                        data.addCardData(cardData);
                        adapter.notifyItemInserted(data.size() - 1);
                    }
                });
                return true;
            case R.id.action_clear:
                data.clearCardData();
                adapter.notifyDataSetChanged();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onCreateContextMenu(@NonNull ContextMenu menu, @NonNull View v, @Nullable ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo);
        requireActivity().getMenuInflater().inflate(R.menu.card_menu, menu);
    }

    @Override
    public boolean onContextItemSelected(@NonNull MenuItem item) {
        int position = adapter.getMenuContextClickPosition();
        switch (item.getItemId()) {
            case R.id.action_update_from_context:
                navigation.addFragment(CardUpdateFragment.newInstance(data.getCardData(position)), true);
                publisher.subscribe(new Observer() {
                    @Override
                    public void updateState(CardData cardData) {
                        data.updateCardData(position, cardData);
                        adapter.notifyItemChanged(position);
                    }
                });
                return true;
            case R.id.action_delete_from_context:
                DialogueDelete dialogueDelete = new DialogueDelete();
                dialogueDelete.show(Objects.requireNonNull(getActivity()).getSupportFragmentManager(),
                        "deletion check");
                if (dialogueDelete.isYes()) {
                    Toast.makeText(getContext(), "deleted", Toast.LENGTH_SHORT).show();
                    //data.deleteCardData(position);
                    //adapter.notifyItemRemoved(position);
                }
                return true;
        }
        return super.onContextItemSelected(item);
    }
}

File: app/src/main/java/ru/geekbrains/notes/view/NotesFragment.kt
package ru.geekbrains.notes.view

import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.view.*
import android.view.ContextMenu.ContextMenuInfo
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import ru.geekbrains.notes.MainActivity
import ru.geekbrains.notes.R
import ru.geekbrains.notes.model.CardData
import ru.geekbrains.notes.model.Note
import ru.geekbrains.notes.presenter.MainPresenter
import ru.geekbrains.notes.presenter.NotesViewContract
import ru.geekbrains.notes.view.CardUpdateFragment.Companion.newInstance
import ru.geekbrains.notes.view.NoteFragment.Companion.newInstance

class NotesFragment : Fragment(), NotesViewContract {
    private val presenter: MainPresenter = MainPresenter(this)
    var currentNote: Note = Note()
    var isLandscape = false
    private var data: List<CardData> = listOf()
    private lateinit var adapter: NotesAdapter
    private lateinit var recyclerView: RecyclerView
    private var navigation: Navigation? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        setHasOptionsMenu(true)
        val view = inflater.inflate(R.layout.fragment_notes, container, false)
        recyclerView = view.findViewById(R.id.recyclerView)
        recyclerView.setHasFixedSize(true)
        val linearLayoutManager = LinearLayoutManager(context)
        recyclerView.layoutManager = linearLayoutManager

        adapter = NotesAdapter(this)
        adapter.setNotesOnClickListener(object : NotesOnClickListener {
            override fun onNoteClick(view: View, position: Int) {
                currentNote = Note(
                    data[position].title,
                    data[position].text
                )
                isLandscape =
                    resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE
                showNote()
            }
        })
        recyclerView.adapter = adapter
        adapter.setDataSource(data)
        presenter.getDataFromSource()

        val defaultItemAnimator = DefaultItemAnimator()
        defaultItemAnimator.changeDuration = 1000
        defaultItemAnimator.removeDuration = 1000
        recyclerView.itemAnimator = defaultItemAnimator
        return view
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        val activity = context as MainActivity
        navigation = activity.navigation
    }

    private fun showNote() {
        if (isLandscape) {
            showNoteLand()
        } else {
            showNotePort()
        }
    }

    private fun showNotePort() {
        requireActivity()
            .supportFragmentManager
            .beginTransaction()
            .replace(R.id.notes_container, newInstance(currentNote))
            .addToBackStack("")
            .commit()
    }

    private fun showNoteLand() {
        requireActivity()
            .supportFragmentManager
            .beginTransaction()
            .replace(R.id.note_container, newInstance(currentNote))
            .commit()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.fragment_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_add -> {
                navigation?.addFragment(newInstance(presenter), true)
                presenter.addCard()
                return true
            }
            R.id.action_clear -> {
                presenter.clear()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
        super.onCreateContextMenu(menu, v, menuInfo)
        requireActivity().menuInflater.inflate(R.menu.card_menu, menu)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val position = adapter.menuContextClickPosition
        when (item.itemId) {
            R.id.action_update_from_context -> {
                navigation?.addFragment(newInstance(data[position], presenter), true)
                presenter.updatePosition(position)
                return true
            }
            R.id.action_delete_from_context -> {
                val dialogueDelete = DialogueDelete(presenter, data[position], position)
                dialogueDelete.show(
                    requireActivity().supportFragmentManager,
                    "deletion check"
                )
                return true
            }
        }
        return super.onContextItemSelected(item)
    }

    override fun showNotes(notes: List<CardData>) {

    }

    override fun setData(data: List<CardData>) {
        this.data = data
    }

    override fun setAdapter(data: List<CardData>) {
        this.adapter.setDataSource(data)
        this.adapter.notifyDataSetChanged()
    }

    companion object {
        fun newInstance(): NotesFragment {
            return NotesFragment()
        }
    }
}
